Use 8k buffer to help clients w/ broken parsing

n1hility opened the pull request:
Introduce an API Handler wrapper that uses a larger buffer (8k) to help legacy clients that expect a full message to appear  in every read(). While not fully solvable on the server side, it will significantly boost their probability of working.

Resolves #15037

```release-note
none
```

[NO NEW TESTS NEEDED]

vrothberg commented on pkg/api/handlers/compat/events.go:
```diff
@@ -63,16 +64,25 @@ func GetEvents(w http.ResponseWriter, r *http.Request) {
-	flush := func() {}
+	// Use a larger buffer to aid some historical docker clients that expect
+	// an event to fit in every read() call
+	o := bufio.NewWriterSize(w, 8192)
```
Docker seems to use WriteFlusher https://github.com/moby/moby/blob/master/api/server/router/system/system_routes.go#L244-L246.

Should we do the same?  It's available in `containers/storege/pkg/ioutils`.

n1hility replied:
So that's the funny thing. Flushing on every write like that should increase the odds of fragmentation. I haven't looked too deeply, but theres a good chance the batched per event behavior is just timing luck (lots of small writes right next to each other that happen to get read together and a pause between events from the added locking that by luck reduces the chance of reading past/across events. Changing to a larger buffer should be more reliable, well as reliable as we can be with this brittle pattern. There could be other factors going on, though.

vrothberg reviewed (commented):
Thanks, @n1hility!

I double checked in the code and confirm Docker is using `net/http`.  I must have followed a wrong path and probably ended up in the grpc code when taking a quick peek.

We should probably consider other endpoints as well.  I see the write flusher being used in a number of endpoints:
```bash
docker (master) $ vgrep NewWriteFlusher
Index File                                      Line Content
    0 pkg/ioutils/writeflusher.go                 83 // NewWriteFlusher returns a new WriteFlusher.
    1 pkg/ioutils/writeflusher.go                 84 func NewWriteFlusher(w io.Writer) *WriteFlusher {
    2 daemon/stats.go                             48 wf := ioutils.NewWriteFlusher(outStream)
    3 api/server/httputils/write_log_stream.go    20 wf := ioutils.NewWriteFlusher(w)
    4 api/server/router/build/build_routes.go    237 output := ioutils.NewWriteFlusher(ww)
    5 api/server/router/plugin/plugin_routes.go  114 output := ioutils.NewWriteFlusher(w)
    6 api/server/router/plugin/plugin_routes.go  149 output := ioutils.NewWriteFlusher(w)
    7 api/server/router/plugin/plugin_routes.go  257 output := ioutils.NewWriteFlusher(w)
    8 api/server/router/system/system_routes.go  244 output := ioutils.NewWriteFlusher(w)
    9 api/server/router/image/image_routes.go     40 output      = ioutils.NewWriteFlusher(w)
   10 api/server/router/image/image_routes.go    121 output := ioutils.NewWriteFlusher(w)
   11 api/server/router/image/image_routes.go    142 output := ioutils.NewWriteFlusher(w)
   12 api/server/router/image/image_routes.go    168 output := ioutils.NewWriteFlusher(w)
```

vrothberg commented:
@Luap99 PTAL

n1hility commented:
> Thanks, @n1hility!
> 
> I double checked in the code and confirm Docker is using `net/http`. I must have followed a wrong path and probably ended up in the grpc code when taking a quick peek.
> 
> We should probably consider other endpoints as well. I see the write flusher being used in a number of endpoints:
> 
> ```shell
> docker (master) $ vgrep NewWriteFlusher
> Index File                                      Line Content
>     0 pkg/ioutils/writeflusher.go                 83 // NewWriteFlusher returns a new WriteFlusher.
>     1 pkg/ioutils/writeflusher.go                 84 func NewWriteFlusher(w io.Writer) *WriteFlusher {
>     2 daemon/stats.go                             48 wf := ioutils.NewWriteFlusher(outStream)
>     3 api/server/httputils/write_log_stream.go    20 wf := ioutils.NewWriteFlusher(w)
>     4 api/server/router/build/build_routes.go    237 output := ioutils.NewWriteFlusher(ww)
>     5 api/server/router/plugin/plugin_routes.go  114 output := ioutils.NewWriteFlusher(w)
>     6 api/server/router/plugin/plugin_routes.go  149 output := ioutils.NewWriteFlusher(w)
>     7 api/server/router/plugin/plugin_routes.go  257 output := ioutils.NewWriteFlusher(w)
>     8 api/server/router/system/system_routes.go  244 output := ioutils.NewWriteFlusher(w)
>     9 api/server/router/image/image_routes.go     40 output      = ioutils.NewWriteFlusher(w)
>    10 api/server/router/image/image_routes.go    121 output := ioutils.NewWriteFlusher(w)
>    11 api/server/router/image/image_routes.go    142 output := ioutils.NewWriteFlusher(w)
>    12 api/server/router/image/image_routes.go    168 output := ioutils.NewWriteFlusher(w)
> ```

Yeah I agree we need to look at the other access points. In things like logs and console output we shouldnt have a problem since clients aren't parsing but anything that could return a mul-record payload is going to need a solution.

rhatdan commented:
Needs tests or  [NO NEW TESTS NEEDED]

n1hility commented:
Updated to do this as part of handler wrapping.

Also updated the following JSON Streaming endpoints that have a similar need as events:
```
/containers/{name}/stats
/containers/{name}/top
/events
/build
```

vrothberg reviewed (commented):
LGTM, thanks @n1hility !

@containers/podman-maintainers PTAL

rhatdan commented:
/approve
/lgtm

n1hility commented:
@vrothberg your welcome and thanks for the review and analysis!
